File: graphs/cpp/is_bipartite.cpp
// Aproach 1 : BFS --> Intution comes from brute force , i.e if a vertex is colored with one color then the adjacent will be having the other color
// make use of queue to put all the nodes whose color have to be checked and conclude finally 

bool isBipartite(vector<vector<int>>& graph) { 
    int V = graph.size() ;
    vector<int>color(V,-1); // all vertices are uncolored initially

    for(int i = 0 ; i < V ; i++){
        if (color[i] == -1){ // start from the first vertex and check 
            if (solve(i,V,graph,color) == false)
                return false;
        }
    }
    return true;
}

bool solve(int u, int V, vector<vector<int>>& graph, vector<int>&color){
    // we make use of queue to put the nodes where adjacent nodes have to be explored
    queue<int>q ; 
    color[u] = 0 ; // initially first node is colored as o 

    q.push(u);
    while(!q.empty()){
        int node = q.front();
        q.pop();

        for(auto x: graph[node]){ // checking adjacent nodes of this node
            if (color[x] == -1){ // if this node is not colored yet
                color[x] = !color[node]; // make sure to assign the opposite color to the currr node
                q.push(x); // push the adjacent onto the queue
            }
            else if (color[x] == color[node]) // if adjacent node has the same color as the curr
                return false; // return false
        }
    }
    // if no such case found with same color for adjacent nodes and all nodes are covered, return true
    return true ; 
}

// TC = O(V+E)
// SC = O(V) // for color array and also for the queue


// Approach2 : DFS --> idea is to simply check for all the uncolored nodes initially and then fill them with 
// a certain color and check the adjacent nodes , just pass on the opposite color to the adjacent nodes 

bool isBipartite(vector<vector<int>>& graph) { 
    int V = graph.size() ;
    vector<int>color(V,-1);

    for(int i = 0 ; i < V ; i++){
        if (color[i] == -1){
            if (dfs(i,0,graph,color) == false)
                return false ;
        }
    }
    return true;
}

bool dfs(int node, int col, vector<vector<int>>&graph, vector<int>&color ){
    color[node] = col ;

    for(auto x: graph[node]){
        if (color[x] == -1){
            if (dfs(x, !col,graph,color) == false)
                return false;
        }
        else if (color[x] == color[node])
            return false;
    }
    return true ;
}

// Time Complexity: O(V + 2E), Where V = Vertices, 2E is for total degrees as we traverse all adjacent nodes.
// Space Complexity: O(3V) ~ O(V), Space for DFS stack space, colour array and an adjacency list.

File: graphs/go/is_bipartite.go
// DFS based approach : Idea is to simply fill the uncolored nodes
// initially with some color and check the color of adjacent nodes
// if same color is found in any of the adjacent nodes , ret false
// else color the adjacent node with the different colors and
// proceed till all the nodes are colored
// TC = O(V+E) , simply for dfs traversal
// SC = O(V) for extra space for colored array

func dfs(node, col int, color []int, graph [][]int) bool {
	color[node] = col

	for _, u := range graph[node] {
		if color[u] == -1 {
			if dfs(u, 1-col, color, graph) == false {
				return false
			}
		} else if color[u] == color[node] {
			return false
		}
	}
	return true
}

func isBipartite(graph [][]int) bool {
	n := len(graph)

	color := make([]int, n)

	for i := 0; i < n; i++ {
		color[i] = -1
	}

	for i := 0; i < n; i++ {
		if color[i] == -1 && dfs(i, 0, color, graph) == false {
			return false
		}
	}
	return true
}

// BFS based approach : just make use of queue to track the
// adjacent nodes and try to find if two adjacent nodes have
// the same color or not , just color all the adjacent nodes of
// a particular node with the opposite color of the node
// procedd and keep checking for the adjacent nodes
// TC = O(V+E) just like for bfs
// SC = O(V) for color array

func bfs(node int, color []int, graph [][]int) bool {
	queue := list.New()
	color[node] = 0 // fill a color to start with

	queue.PushBack(node)
	//fmt.println(queue.Front().Value)

	for queue.Len() != 0 {
		first := queue.Front().Value.(int)
		queue.Remove(queue.Front())

		for _, v := range graph[first] {
			if color[v] == -1 {
				color[v] = 1 - color[first]
				queue.PushBack(v)
			} else if color[v] == color[first] {
				return false
			}
		}
	}
	return true
}

func isBipartite(graph [][]int) bool {
	n := len(graph)

	color := make([]int, n)

	for i := 0; i < n; i++ {
		color[i] = -1
	}

	for i := 0; i < n; i++ {
		if color[i] == -1 && bfs(i, color, graph) == false {
			return false
		}
	}
	return true
}
